### PR TITLE
Create Google Storage integration tests.

### PR DESCRIPTION
**Pull request by dsuhinin**

### CODE REVIEW

File: pkg/api/mlflow/service/artifact/storage/gs.go [hunk 1]
```diff
@@ -25,7 +27,19 @@ type GS struct {
 }
 
 // NewGS creates new GC instance.
-func NewGS(ctx context.Context, _ *config.ServiceConfig) (*GS, error) {
+func NewGS(ctx context.Context, config *config.ServiceConfig) (*GS, error) {
+	if config.GSEndpointURI != "" {
+		client, err := storage.NewClient(
+			ctx, option.WithEndpoint(config.GSEndpointURI), option.WithoutAuthentication(),
```
Comment (dsuhinin) [inline pkg/api/mlflow/service/artifact/storage/gs.go:33]: @jgiannuzzi the only one concern I can see here -> when we use custom endpoint we need an option to disable `authentication`. So we have 2 ways:
- just hardcode it as is - because only the case when we need to use `custom` endpoint is `tests`. nobody else will do that in any other case.
- we can of course make it configurable via additional flag. so for example -> `FML_GS_DISABLE_AUTH` so in that case everything  will be super configurable.
what do you think? My choose is 1 =) until we really need it.

Comment (jgiannuzzi) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: I agree with you on 1 (+ some comments explaining why it's hardcoded)
But here is another idea: let's reuse our `config.devMode` setting for enabling `option.WithoutAuthentication()`!

File: pkg/api/mlflow/config/config.go [hunk 2]
```diff
@@ -35,6 +36,7 @@ func NewServiceConfig() *ServiceConfig {
 		AuthPassword:          viper.GetString("auth-password"),
 		DefaultArtifactRoot:   viper.GetString("default-artifact-root"),
 		S3EndpointURI:         viper.GetString("s3-endpoint-uri"),
+		GSEndpointURI:         viper.GetString("gs-endpoint-uri"),
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/config/config.go:39]: please hide this setting, as users won't actually need it

File: pkg/api/mlflow/service/artifact/storage/gs.go (hunk 1 above)
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: hm, ok!. let me quickly fix it!

File: pkg/api/mlflow/service/artifact/storage/gs.go [hunk 3]
```diff
@@ -25,10 +27,22 @@ type GS struct {
...
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/service/artifact/storage/gs.go:33]: we should build a list of options here instead, and then pass it `storage.NewClient` below, so it's always called in the same place — your current approach duplicates quite a bit of code

File: pkg/api/mlflow/service/artifact/storage/gs.go [hunk 4]
```diff
@@ -97,6 +111,7 @@ func (s GS) Get(ctx context.Context, artifactURI, path string) (io.ReadCloser, e
 	// 2. get object from gcp storage.
 	reader, err := s.client.Bucket(bucketName).Object(filepath.Join(prefix, path)).NewReader(ctx)
 	if err != nil {
+		fmt.Println("err:", err)
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/service/artifact/storage/gs.go:114]: looks like a remnant of some debugging

File: tests/integration/docker-compose.yml [hunk 5]
```diff
@@ -49,6 +52,17 @@ services:
...
 
+  google-storage:
+    image: fsouza/fake-gcs-server
+    ports:
+      - 4443:4443
+    entrypoint: sh
+    command: >
+      -c 'mkdir -p /storage/bucket1 && 
+          mkdir -p /storage/bucket2 && 
+          mkdir -p /storage/bucket3 &&
```
Comment (jgiannuzzi) [inline tests/integration/docker-compose.yml:63]: buckets should be created/deleted by the tests themselves

File: tests/integration/golang/mlflow/artifact/get_artifact_s3_test.go [hunk 6]
```diff
@@ -102,10 +102,7 @@ func (s *GetArtifactS3TestSuite) Test_Ok() {
 
 			// upload artifact subdir object to S3
 			putObjReq = &s3.PutObjectInput{
-				Key: aws.String(fmt.Sprintf(
-					"1/%s/artifacts/artifact.subdir/artifact.file",
-					runID),
-				),
+				Key:    aws.String(fmt.Sprintf("1/%s/artifacts/artifact.subdir/artifact.file", runID)),
```
Comment (jgiannuzzi) [inline tests/integration/golang/mlflow/artifact/get_artifact_s3_test.go:105]: this change is unrelated _and_ makes this less readable (IMHO)

File: tests/integration/golang/helpers/env.go [hunk 7]
```diff
@@ -26,6 +26,14 @@ func GetS3EndpointUri() string {
 	return "http://localhost:9000"
 }
 
+func GetGSEndpointUri() string {
+	uri, ok := os.LookupEnv("FML_GS_ENDPOINT_URI")
+	if ok {
+		return uri
+	}
+	return "http://localhost:4333"
```
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/env.go:34]: this is nice but we then need to have the fake GCS server brought up with the devcontainer

File: pkg/api/mlflow/service/artifact/storage/gs.go (hunk 4 above)
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:114]: damn. yep. removed everything else but no this. done.

File: pkg/api/mlflow/config/config.go (hunk 2 above)
Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:39]: aha, I see your point. done.

File: pkg/api/mlflow/service/artifact/storage/gs.go (hunk 3 above)
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: yes, you are right. done.

File: pkg/api/mlflow/service/artifact/storage/gs.go (hunk 1 above)
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: oook. have to revert =)

Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: done.

File: tests/integration/golang/mlflow/artifact/get_artifact_s3_test.go (hunk 6 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/artifact/get_artifact_s3_test.go:105]: yep. ok.

Comment (dsuhinin) [reply tests/integration/golang/mlflow/artifact/get_artifact_s3_test.go:105]: done.

File: tests/integration/docker-compose.yml (hunk 5 above)
Comment (dsuhinin) [reply tests/integration/docker-compose.yml:63]: oh, let me try. this `fake` guys almost killed me because of stupid configuration so I really don't want to touch anything but let me try of course.

Comment (dsuhinin) [reply tests/integration/docker-compose.yml:63]: done.

File: pkg/api/mlflow/config/config.go [hunk 8]
```diff
@@ -19,6 +19,7 @@ type ServiceConfig struct {
 	AuthPassword          string
 	DefaultArtifactRoot   string
 	S3EndpointURI         string
+	GSEndpointURI         string
```
Comment (suprjinx) [inline pkg/api/mlflow/config/config.go:22]: I wonder if "GS" is too short to be clear? "S3" is okay because everyone knows it, but "GoogleStorage" might be clearer here. Or comment that "this is Google Storage"

File: pkg/api/mlflow/service/artifact/storage/gs.go [hunk 9]
```diff
@@ -25,10 +26,16 @@ type GS struct {
 }
 
 // NewGS creates new GC instance.
```
Comment (suprjinx) [inline pkg/api/mlflow/service/artifact/storage/gs.go:28]: Abbreviations too brief?

File: tests/integration/golang/helpers/env.go (hunk 7 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/env.go:34]: done

File: pkg/api/mlflow/service/artifact/storage/gs.go (hunk 9 above)
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go]: yep, changed comments to use `Google Storage` instead of `GS`.

File: pkg/api/mlflow/config/config.go (hunk 8 above)
Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:22]: it looks consistent with `S3` and this parameter really needs only for integration tests. It is hidden even from user eyes

**Review by jgiannuzzi — COMMENTED** [review]
Oops, I forgot to submit all of these pending comments from yesterday!

File: tests/integration/golang/helpers/gs.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package helpers
```
[the remark this answers is not in the document]
Comment (dsuhinin) [reply tests/integration/golang/helpers/gs.go:1]: probably  we do not use them in main code, so the shouldn't be included into final build. I can add `build` tag just to make sure.

File: pkg/api/mlflow/service/artifact/storage/gs.go [hunk 11]
```diff
@@ -28,19 +27,13 @@ type GS struct {
...
-			ctx, option.WithEndpoint(config.GSEndpointURI), option.WithoutAuthentication(),
-		)
-		if err != nil {
-			return nil, eris.Wrap(err, "error creating GS storage client")
-		}
-		return &GS{
-			client: client,
-		}, nil
+		// include option.WithoutAuthentication() option, because otherwise standard GCP DSK won't work properly.
+		// make it configurable via ENV if it's really needed.
```
[the remark this answers is not in the document]
Comment (dsuhinin) [reply pkg/api/mlflow/service/artifact/storage/gs.go:33]: done.

File: tests/integration/golang/mlflow/artifact/get_artifact_gs_test.go [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+		{
+			name:  "IncorrectPathProvidedCase5",
+			error: api.NewInvalidParameterValueError("provided 'path' parameter is invalid"),
+			request: request.GetArtifactRequest{
+				RunID: "run_id",
+				Path:  "/foo/../bar",
+			},
+		},
+		{
+			name: "S3IncompletePath",
```
[the remark this answers is not in the document]
Comment (dsuhinin) [reply tests/integration/golang/mlflow/artifact/get_artifact_gs_test.go]: done.

File: .devcontainer/docker-compose.yml [hunk 13]
```diff
@@ -78,3 +79,8 @@ services:
       MINIO_CONSOLE_ADDRESS: ":9001"
     command: server /data
     network_mode: service:db
+
+  google-storage:
+    image: fsouza/fake-gcs-server
+    entrypoint: sh
+    command: -c 'exec fake-gcs-server -public-host=google-storage -scheme http'
```
Comment (jgiannuzzi) [inline .devcontainer/docker-compose.yml:86]:
```suggestion
    command:
      - -public-host=google-storage
      - -scheme=http
```

File: tests/integration/docker-compose.yml [hunk 14]
```diff
@@ -53,14 +53,8 @@ services:
...
     image: fsouza/fake-gcs-server
-    ports:
-      - 4443:4443
     entrypoint: sh
-    command: >
-      -c 'mkdir -p /storage/bucket1 && 
-          mkdir -p /storage/bucket2 && 
-          mkdir -p /storage/bucket3 && 
-          exec fake-gcs-server -public-host=google-storage -scheme http'
+    command: -c 'exec fake-gcs-server -public-host=google-storage -scheme http'
```
Comment (jgiannuzzi) [inline tests/integration/docker-compose.yml:57]: same remark as above

File: tests/integration/golang/helpers/gs.go [hunk 15]
```diff
@@ -19,3 +24,31 @@ func NewGSClient(endpoint string) (*storage.Client, error) {
 
 	return client, nil
 }
+
+// PrepareTestBuckets prepares test buckets for further usage in tests:
+// - delete provided buckets and all the content within them.
+// - create the same empty bucket.
+func PrepareTestBuckets(client *storage.Client, buckets []string) error {
```
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/gs.go:31]: I think this should be split between delete (called during teardown) and create (called during setup)

File: tests/integration/golang/helpers/gs.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply tests/integration/golang/helpers/gs.go:1]: but is it intentional that it's only for helpers and fixtures? (genuine question)

File: .devcontainer/docker-compose.yml (hunk 13 above)
Comment (dsuhinin) [reply .devcontainer/docker-compose.yml:86]: ok.

File: tests/integration/docker-compose.yml (hunk 14 above)
Comment (dsuhinin) [reply tests/integration/docker-compose.yml:57]: ok

File: tests/integration/golang/helpers/gs.go (hunk 15 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/gs.go]: done.

File: tests/integration/golang/helpers/gs.go [hunk 16]
```diff
@@ -25,10 +25,18 @@ func NewGSClient(endpoint string) (*storage.Client, error) {
 	return client, nil
 }
 
-// PrepareTestBuckets prepares test buckets for further usage in tests:
-// - delete provided buckets and all the content within them.
-// - create the same empty bucket.
-func PrepareTestBuckets(client *storage.Client, buckets []string) error {
+// CreateGSBuckets creates tests buckets in the Google Storage.
+func CreateGSBuckets(client *storage.Client, buckets []string) error {
```
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/gs.go:29]: I really like that you're passing the buckets here instead of having some global variable — could you please modify the S3 tests to do the same thing? (in another PR, of course)

File: tests/integration/golang/helpers/env.go (hunk 7 above)
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/env.go:34]: I see that `FML_GS_ENDPOINT_URI` in the dev container points to `http://localhost:4443/storage/v1/` — which one is the good one?

File: tests/integration/golang/helpers/gs.go (hunk 16 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/gs.go:29]: sure!

File: tests/integration/golang/helpers/env.go (hunk 7 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/env.go]: yep, this one is correct -> http://localhost:4443/storage/v1/. fixed.

File: tests/integration/golang/helpers/gs.go (hunk 16 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/gs.go:29]: @jgiannuzzi addressed #509

File: .devcontainer/devcontainer.json [hunk 17]
```diff
@@ -32,6 +32,7 @@
 	"forwardPorts": [
 		5000,
 		5432,
+		4333,
```
Comment (jgiannuzzi) [inline .devcontainer/devcontainer.json:35]: wrong port

File: .devcontainer/docker-compose.yml [hunk 18]
```diff
@@ -78,3 +79,9 @@ services:
       MINIO_CONSOLE_ADDRESS: ":9001"
     command: server /data
     network_mode: service:db
+
+  google-storage:
+    image: fsouza/fake-gcs-server
+    command:
+      -public-host=google-storage
+      -scheme=http
```
Comment (jgiannuzzi) [inline .devcontainer/docker-compose.yml:87]: missing network_mode

File: tests/integration/golang/helpers/env.go [hunk 19]
```diff
@@ -26,6 +26,14 @@ func GetS3EndpointUri() string {
 	return "http://localhost:9000"
 }
 
+func GetGSEndpointUri() string {
+	uri, ok := os.LookupEnv("FML_GS_ENDPOINT_URI")
+	if ok {
+		return uri
+	}
+	return "http://localhost:4333/storage/v1/"
```
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/env.go:34]: wrong port

File: .devcontainer/devcontainer.json (hunk 17 above)
Comment (dsuhinin) [reply .devcontainer/devcontainer.json]: fixed.

File: .devcontainer/docker-compose.yml (hunk 18 above)
Comment (dsuhinin) [reply .devcontainer/docker-compose.yml:87]: yep. added already.

File: tests/integration/golang/helpers/env.go (hunk 19 above)
Comment (dsuhinin) [reply tests/integration/golang/helpers/env.go]: yep, already fixed.

File: tests/integration/docker-compose.yml [hunk 20]
```diff
@@ -48,6 +51,12 @@ services:
       MINIO_ROOT_PASSWORD: "password"
       MINIO_CONSOLE_ADDRESS: ":9001"
 
+  google-storage:
```
[the remark this answers is not in the document]
Comment (dsuhinin) [reply tests/integration/docker-compose.yml:54]: I tried to add it, but probably `curl` is not exists within this container so I can't ping it.

**Review by suprjinx — DISMISSED** [review]
LGTM

**Review by jgiannuzzi — CHANGES_REQUESTED** [review]
the integration tests don't work when I run them from the dev container (so starting the server with "Launch server (supports integration tests)" and then running the integration tests from the Test Explorer)
I also noticed that the list tests seem to be using the S3 storage instead of GS!
```
DEBU[0209]/workspaces/fasttrackml/pkg/api/mlflow/service/artifact/storage/s3.go:91 github.com/G-Research/fasttrackml/pkg/api/mlflow/service/artifact/storage.S3.List() got 0 directories from S3 storage for bucket "bucket2" and prefix "1/63fbeb4e81be4a20b0c27fa893ab8d0f/artifacts/non-existing-dir/" 
DEBU[0209]/workspaces/fasttrackml/pkg/api/mlflow/service/artifact/storage/s3.go:104 github.com/G-Research/fasttrackml/pkg/api/mlflow/service/artifact/storage.S3.List() got 0 objects from S3 storage for bucket "bucket2" and prefix "1/63fbeb4e81be4a20b0c27fa893ab8d0f/artifacts/non-existing-dir/" 
```

File: tests/integration/docker-compose.yml [hunk 21]
```diff
@@ -56,6 +56,10 @@ services:
     command:
       -public-host=google-storage
       -scheme=http
+    healthcheck:
+      test: [ "CMD", "wget", "-nv -t1 -spider", "http://google-storage:4443/storage/v1/b" ]
```
Comment (jgiannuzzi) [inline tests/integration/docker-compose.yml:60]:
```suggestion
      test: [ "CMD", "wget", "-q", "--spider", "http://localhost:4443/storage/v1/b" ]
```

Comment (dsuhinin) [reply tests/integration/docker-compose.yml:60]: ahh, ok!

File: .devcontainer/docker-compose.yml [hunk 22]
```diff
@@ -83,6 +83,6 @@ services:
   google-storage:
```
Comment (jgiannuzzi) [inline .devcontainer/docker-compose.yml:83]: please add a volume like we do for the other service containers (everything is stored in the `/storage` directory)

Comment (dsuhinin) [reply .devcontainer/docker-compose.yml:84]: done.

File: .devcontainer/docker-compose.yml [hunk 23]
```diff
@@ -78,3 +80,12 @@ services:
...
     command: server /data
     network_mode: service:db
+
+  google-storage:
+    image: fsouza/fake-gcs-server
+    volumes:
+      - google-storage-data:/storage
+    command:
+      -public-host=localhost
+      -scheme=http
```
Comment (jgiannuzzi) [inline .devcontainer/docker-compose.yml:90]:
```suggestion
    command:
      - -public-host=localhost
      - -scheme=http
```

File: tests/integration/docker-compose.yml [hunk 24]
```diff
@@ -48,6 +51,16 @@ services:
       MINIO_ROOT_PASSWORD: "password"
       MINIO_CONSOLE_ADDRESS: ":9001"
 
+  google-storage:
+    image: fsouza/fake-gcs-server
+    command:
+      -public-host=google-storage
+      -scheme=http
```
Comment (jgiannuzzi) [inline tests/integration/docker-compose.yml:58]:
```suggestion
      - -public-host=google-storage
      - -scheme=http
```

File: .devcontainer/docker-compose.yml (hunk 23 above)
Comment (dsuhinin) [reply .devcontainer/docker-compose.yml]: done.

File: tests/integration/docker-compose.yml (hunk 24 above)
Comment (dsuhinin) [reply tests/integration/docker-compose.yml]: done.

File: tests/integration/golang/helpers/s3.go [hunk 25]
```diff
@@ -34,9 +34,9 @@ func NewS3Client(endpoint string) (*s3.Client, error) {
 	}), nil
 }
 
-// CreateS3Buckets creates the test buckets.
-func CreateS3Buckets(s3Client *s3.Client) error {
-	for _, bucket := range testBuckets {
+// CreateBuckets creates the test bucekts.
```
Comment (jgiannuzzi) [inline tests/integration/golang/helpers/s3.go:37]:
```suggestion
// CreateBuckets creates the test buckets.
```

Comment (dsuhinin) [reply tests/integration/golang/helpers/s3.go]: done

**Review by jgiannuzzi — APPROVED** [review]
